File: Easy/maximum_subarray.py
####### My Solution ####### 88.3%

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        for i in range(1, len(nums)):
            if nums[i-1] > 0:
                nums[i] += nums[i-1]

        return max(nums)


### Divide and Conquer
class Solution2:
    def maxSubArray(self, nums: List[int]) -> int:


        def helper(nums,left,right):
            mid = (left + right) // 2
            curr = best_left_sum = best_right_sum = 0
            for i in range(mid,left,-1):
                curr += nums[i]
                best_left_sum = max(best_left_sum,curr)

            curr = 0

            for i in range(mid,right):
                curr += nums[i]
                best_right_sum = max(best_right_sum,curr)
            
            
            

File: Easy/maximum_subarray.c
/*  Online Solution  */
int maxSubArray(int* nums, int numsSize){
    int i, tmp, max;
    for ( i = 0, tmp = 0, max = nums[0]; i < numsSize; i++)
    {
        tmp += nums[i];

        if (tmp>max)
        {
            max = tmp;
        }

        if (tmp < 0)
        {
            tmp = 0;
        }
        
        
        
    }
    return  max;
}

/*  Online Solution  */
static int midCrossMax(int* arr, int size)
{
    int lmax, rmax, tmp, i, m = (size / 2) - 1;
    
    for (i = m, tmp = 0, lmax = arr[m]; i >= 0; i--) {
        tmp += arr[i];
        lmax = tmp > lmax ? tmp : lmax;
    }
    
    for (i = m + 1, tmp = 0, rmax = arr[m + 1]; i < size; i++) {
        tmp += arr[i];
        rmax = tmp > rmax ? tmp : rmax;
    }

    return lmax + rmax;
}

int maxSubArray(int* nums, int numsSize)
{
    int m = numsSize / 2;
    int max, l, r, c;
    
    if (numsSize == 1)
        return *nums;
    
    l = maxSubArray(nums, m);
    r = maxSubArray(nums + m, numsSize - m);
    c = midCrossMax(nums, numsSize);
    
    max = l > r ? l : r;
    max = max > c ? max : c;

    return max;
    
}